pkg/rebuild/maven/infer_test.go

```diff
@@ -364,3 +364,150 @@ func must[T any](t T, err error) T {
 	}
 	return t
 }
+
+func TestGitIndexScan(t *testing.T) {
+	testCases := []struct {
+		name                  string
+		repoYAML              string
+		zipEntries            []*archive.ZipEntry
+		expectedCommitMessage string
+		expectedError         bool
+	}{
+		{
+			name: "simple case with direct match",
+			repoYAML: `
+            commits:
+              - id: initial-commit
+                message: initial-commit
+                files:
+                  src/main/java/com/example/App.java: |
+                    a`,
+			zipEntries: []*archive.ZipEntry{
+				{
+					FileHeader: &zip.FileHeader{Name: "com/example/App.java"},
+					Body:       []byte("a"),
+				},
+			},
+			expectedCommitMessage: "initial-commit",
+		},
+		{
+			name: "parent match",
+			repoYAML: `
+            commits:
+              - id: initial-commit
+                message: initial-commit
+                files:
+                  src/main/java/com/example/App.java: |
+                    a
+              - id: second-commit
+                message: second-commit
+                files:
+                  src/main/java/com/example/App.java: |
+                    b`,
+			zipEntries: []*archive.ZipEntry{
+				{
+					FileHeader: &zip.FileHeader{Name: "com/example/App.java"},
+					Body:       []byte("b"),
+				},
+			},
+			expectedCommitMessage: "second-commit",
```

Comment:
nit: you should be able to check against the commit ID directly (using the RepoMap) and remove it here.